Add documentation for CLI architecture (#12093)
pypa/pip: docs/html/development/architecture/command-line-interface.rst

``Command`` then uses another ``ConfigOptionParser`` instance, to parse command-specific args.

Command structure
-----------------

This section shows the class hierarchy from which every command's class will inherit
from.

`base_command.py <https://github.com/pypa/pip/blob/main/src/pip/_internal/cli/base_command.py>`_
defines the base ``Command`` class, from which every other command will inherit directly or
indirectly (see the *command tree* at the end of this section).

Using the ``ConfigOptionParser`` (see `Configuration and CLI "blend" <Configuration and CLI "blend"_>`_),
this class adds the general options and instantiates the *cmd_opts* group, where every other specific
option will be added if needed on each command's class. For those commands that define specific
options, like ``--dry-run`` on ``pip install`` command, the options must be added to *cmd_opts*
this is the job of *add_options* method), which will be automatically called on ``Command``'s initialization.

The base ``Command`` has the following methods:

.. py:class:: Command

  .. py:method:: main()

>     Main method of the class, its always called (as can be seen in main.py's

Reviewer: ```suggestion
    Main method of the class, it's always called (as can be seen in main.py's
```

    `main <https://github.com/pypa/pip/blob/main/src/pip/_internal/cli/main.py#L46>`_).
    It's in charge of calling the specific ``run`` method of the class and handling the possible errors.

  .. py:method:: run()

    Abstract method where the actual action of a command is defined.

  .. py:method:: add_options()

    Optional method to insert additional options on a class, called on ``Command`` initialization.

Some commands have more specialized behavior, (see for example ``pip index``).
These commands instead will inherit from ``IndexGroupCommand``, which inherits from ``Command``
and  ``SessionCommandMixin`` to build build the pip session for the corresponding requests.

Lastly, ``RequirementCommand``, which inherits from ``IndexGroupCommand`` is the base class
for those commands which make use of requirements in any form, like ``pip install``.

In addition to the previous classes, a last mixin class must be mentioned, from which
``Command`` as well as ``SessionCommandMixin`` inherit: ``CommandContextMixIn``, in
charge of the command's context.

> In the following command tree we can see the hierarchy defined for the different pip's
> commands, where each command is defined under the base class it inherits from:

Reviewer: ```suggestion
In the following command tree we can see the hierarchy defined for the different pip
commands, where each command is defined under the base class it inherits from:
```

| ``Command``
|  ├─ ``cache``, ``check``, ``completion``, ``configuration``, ``debug``, ``freeze``, ``hash``, ``help``, ``inspect``, ``show``, ``search``, ``uninstall``
|  └─ ``IndexGroupCommand``
|      ├─ ``index``, ``list``
|      └─ ``RequirementCommand``

> |           └─ ``wheel``, ``download``, ``install``

Reviewer 2: How does this render?

Author: ![image](https://github.com/pypa/pip/assets/56895847/80a37300-71b4-49c8-bc04-4ce1ce2f1bab)

I thought it may help, but if its ugly or doesn't help at all, I'll remove it.

Reviewer 2: I was only worrying it might not render well. Seems fine though so good for me.

Author: Hi @uranusjr if there is anything else I should do to add this changes, please let me know.

command using the information stored in `commands_dict <https://github.com/pypa/pip/blob/main/src/pip/_internal/commands/__init__.py>`_
variable, and calling its ``main`` method (see `Command structure <Command structure>`_).

A second argument parsing is done at each specific command (defined in the base ``Command`` class),
again using the ``ConfigOptionParser``.

Argument access
---------------

To access all the options and arguments, ``Command.run()`` takes
the options as `optparse.Values <https://docs.python.org/3/library/optparse.html#optparse.Values>`_
and a list of strings for the arguments (parsed in ``Command.main()``). The internal methods of
the base ``Command`` class are in charge of passing these variables after ``parse_args`` is
called for a specific command.

Configuration and CLI "blend"
-----------------------------

The base ``Command`` instantiates the class `ConfigOptionParser <https://github.com/pypa/pip/blob/main/src/pip/_internal/cli/parser.py>`_
which is in charge of the parsing process (via its parent class
`optparse.OptionParser <https://docs.python.org/3/library/optparse.html#optparse.OptionParser>`_).

> Its main addition consists on the following function:

Reviewer: ```suggestion
Its main addition consists of the following function:
```

.. py:class:: ConfigOptionParser(OptionParser)

  .. py:method:: get_default_values()

    Overrides the original method to allow updating the defaults ater the instantiation of the
    option parser.

It allows overriding the default options and arguments using the ``Configuration`` class

> (more information can be found on :ref:`Configuration`) to add environment variables and
> information from configuration files.

Reviewer: ```suggestion
(more information can be found on :ref:`Configuration`) to include environment variables and
settings from configuration files.
```

Progress bars and spinners
--------------------------

> There are two more modules in the cli subpackage in charge of showing the state of the

Reviewer: ```suggestion
There are two more modules in the ``cli`` subpackage in charge of showing the state of the
```
